### PR TITLE
feat(search): HNSW

### PR DESCRIPTION
**Pull request by dranikpg**

### CODE REVIEW

File: src/core/search/indices.cc [hunk 1]
```diff
@@ -121,16 +135,97 @@ void VectorIndex::Add(DocId id, DocumentAccessor* doc, string_view field) {
...
+      }
+      const vector<DocId>* allowed;
+    };
+
+    BinsearchFilter filter{&allowed};
+    return QueueToVec(world_.searchKnn(target, k, &filter));
+  }
+
+ private:
+  using SpaceUnion = std::variant<hnswlib::L2Space, hnswlib::InnerProductSpace>;
```
Comment (chakaz) [inline src/core/search/indices.cc:179]: I **love** the name :laughing:

File: src/core/search/search_parser_test.cc [hunk 2]
```diff
@@ -12,9 +12,9 @@ namespace dfly::search {
 
 using namespace std;
 
-class SearchParserTest : public ::testing::Test {
+class SearchTest : public ::testing::Test {
```
Comment (chakaz) [inline src/core/search/search_parser_test.cc:15]: I get why you changed the test names in `search_test.cc`, but did you intend to rename here as well?

File: src/server/search/search_family.cc [hunk 3]
```diff
@@ -46,28 +46,38 @@ bool IsValidJsonPath(string_view path) {
...
 
-pair<size_t, search::VectorSimilarity> ParseVectorFieldInfo(CmdArgParser* parser,
-                                                            ConnectionContext* cntx) {
+search::SchemaField::VectorParams ParseVectorParams(CmdArgParser* parser) {
   size_t dim = 0;
-  search::VectorSimilarity sim = search::VectorSimilarity::L2;
+  auto sim = search::VectorSimilarity::L2;
+  size_t capacity = 1000;
 
+  bool use_hnsw = parser->ToUpper().Next().Case("HNSW", true).Case("FLAT", false);
```
Comment (chakaz) [inline src/server/search/search_family.cc:54]: :exploding_head:

**Review by chakaz — APPROVED** [review]

File: src/core/search/search_parser_test.cc (hunk 2 above)
Comment (dranikpg) [reply src/core/search/search_parser_test.cc:15]: Ops, I didn't, vscode did its job to well

File: src/server/search/search_family.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/search/search_family.cc:54]: Better than 5+ lines `if {} else if {} else reply` 😼

Comment (chakaz) [reply src/server/search/search_family.cc:54]: Don't get me wrong, I'm not saying it's bad. It's great, readable, no complaints.
I was just blown away that this is C++ and not, idk, Python :rofl:

**Review by chakaz — APPROVED** [review]

File: src/server/search/search_family.cc (hunk 3 above)
Comment (kostasrim) [reply src/server/search/search_family.cc:54]: > Don't get me wrong, I'm not saying it's bad. It's great, readable, no complaints.
I was just blown away that this is C++ and not, idk, Python

How do you feel about `std::ranges` then ? :stuck_out_tongue:

**Pull request merged**